Add comments/docstrings to exceptions

craddm opened the pull request:
## :white_check_mark: Checklist

<!--
Thank you for your pull request!

Before going any further, please review the [guidelines for contributing](../CONTRIBUTING.md) to this repository.
Replace the empty checkboxes [ ] below with checked ones [x] accordingly.
-->

- [x] You have given your pull request a meaningful title (_e.g._ `Enable foobar integration` rather than `515 foobar`).
- [x] You are targeting the appropriate branch. If you're not certain which one this is, it should be **`develop`**.
- [x] Your branch is up-to-date with the **target branch** (it probably was when you started, but it may have changed since then).
- [x] You have marked this pull request as a **draft** and added `'[WIP]'` to the title if needed (if you're not yet ready to merge).

### :arrow_heading_up: Summary

<!--
Please explain what your pull request does here.
You might (optionally) want to include screenshots here.
-->

Adds descriptive comments and/or docstrings to exceptions.

Also checks consistency of usage across the codebase.

### :closed_umbrella: Related issues

<!--
If your pull request will close any open issues (hopefully it will!) then add `Closes #<issue number>` here.
-->

Closes #1518

### :microscope: Tests

<!--
- Document any manual tests that you have carried out (*e.g.* deploying a new SHM and/or SRE) and confirm which commit you did this with
- Note that automated tests will be run as part of the CI process and will block this PR until they pass.
- Additionally, a successful code review may be required before this PR can be merged.
- If this pull request only changed documentation you can simply state 'Documentation only' here.
-->

craddm committed:
Add comment for EntraID error

jemrobinson commented on data_safe_haven/exceptions/__init__.py:
```diff
@@ -9,7 +9,8 @@ class DataSafeHavenCloudError(DataSafeHavenError):
     pass
 
-
+# This is a custom exception class for handling errors when interacting with Entra ID.
+# For example, when adding users to an Entra group fails.
 class DataSafeHavenEntraIDError(DataSafeHavenCloudError):
     pass
```
```suggestion
class DataSafeHavenEntraIDError(DataSafeHavenCloudError):
    """
    This is a custom exception class for handling errors when interacting with Entra ID.
    For example, when adding users to an Entra group fails.
    """
    pass
```

If you make this a docstring instead of a comment then it will show up on e.g. tooltips in an IDE.

JimMadge replied:
And in API documentation if we ever do that.

craddm committed:
Move comment to docstring

craddm committed:
Add docstring for configerror

craddm committed:
Add docstring for SSLError

craddm committed:
shorten descriptive text in docstrings

craddm committed:
Add docstring for CloudError

craddm committed:
Add docstring for CloudError

craddm committed:
Add docstring for IPRangeError

craddm committed:
Finish AzureError docstring, add for DataSafeHavenError

craddm committed:
Add docstring for DataSafeHavenError

craddm committed:
SSLError docstring

craddm commented:
Currently the custom exceptions use the following hierarchy: 
![image](https://github.com/alan-turing-institute/data-safe-haven/assets/5796417/8b1195f0-4c49-489e-9ecf-9d4479964773)

craddm committed:
Merge branch 'alan-turing-institute:develop' into add-exception-comments

craddm committed:
add .venv/ to gitignore

craddm committed:
add several further expection docstrings

craddm committed:
Add additional exception docstrings

craddm committed:
document use of DSHInputError

craddm committed:
Use Pulumi error for consistency

craddm commented:
I'm still trying to understand what we want the errors to achieve. At the moment they're a bit of a mishmash. 

For example, `DataSafeHavenParameterError` is used when a user supplies a key for a `context` that either does not exist or already exists (when trying to create), but also when a yaml config file cannot be validated.

On the other hand, we have `DataSafeHavenConfigError`, which is used when no context file exists, or when a user supplies an SRE name that does not exist in the config, but also when no context is selected, or when the config file is not correct yaml.

Then we have `DataSafeHavenInputError` which is used when a subscription cannot be found, when there are missing values for users, when a secret already exists in a key-vault...

Finding it difficult to nail a description for some of these error classes as it's not always clear what they are intended to capture

jemrobinson commented:
It's totally fine for you to redesign the error names and/or inheritance structure to make more sense. I think we want things to adhere to the following guidelines:

- everything should inherit from `DataSafeHavenError` eventually
- if `X` and `Y` both inherit from `A` it should be because they're conceptually both a type of `A` but that it's worth distinguishing between them (e.g. it might be nice to distinguish Azure errors from Graph errors even though they're both some kind of Microsoft/Cloud error).
- we (ideally) should be using each type of error in more than one place and, if possible, more than one file

Anything else you'd add to this @JimMadge ?

jemrobinson commented:
@craddm is this still being worked on?

craddm commented:
Yes - sorry, it's been bumped down the priority list a few times. I'll get back to it shortly.

jemrobinson commented:
No problem - just checking that you weren't waiting for review.

jemrobinson commented:
Um, what happened here @craddm ?

craddm commented:
I tried to update it inline with develop, and apparently it went horribly wrong

jemrobinson committed:
Merge branch 'develop' into add-exception-comments

craddm committed:
Merge branch 'alan-turing-institute:develop' into add-exception-comments

craddm committed:
Add more detailed comments and notes for error classes

craddm committed:
Remove DataSafeHavenNotImplementedError

craddm committed:
Merge branch 'develop' into add-exception-comments

craddm committed:
Remove unnecessary spaces

craddm committed:
Reorder error classes

craddm committed:
Grammar fixes

craddm committed:
More concise docstring for IPRangeError

craddm committed:
Make userhandlingerror no longer an internalerror

craddm committed:
remove internalerror from azure_api and replace with AzureError

craddm committed:
Add DSHType andDSHValue Error classes

craddm committed:
change parameter errors to valueerrors

craddm committed:
Change DSHInputErrors to DSHTypeErrors for Entra users

craddm committed:
Change from InputError to ValueError

craddm committed:
Change from InputError to ValueError

craddm committed:
Change from InputError to ValueError

craddm committed:
Remove DSHInputError class

craddm committed:
Replace InputError with ValueError

craddm committed:
Replace DSHInputError with DSHPulumiError

craddm committed:
Fix linting errors

craddm committed:
Replace DSHInternal with DSHMicrosoftGraph errors

craddm committed:
Replace ParameterError with TypeError

craddm committed:
fix linting error

craddm committed:
Update strings tests to check for Value rather than Input errors

craddm committed:
Remove DSHInternalError class

craddm committed:
Update tests from ParameterError to TypeError

craddm committed:
Update tests to reflect changes in error types

craddm committed:
Change ParameterError to TypeError

craddm committed:
Change ParameterError to Type and ValueErrors as applicable

craddm committed:
Change to DSHTypeError

craddm committed:
Remove DSHParameterError class

craddm committed:
Change tests to check for TypeError

craddm committed:
Decode bytes for printing

craddm committed:
Merge branch 'alan-turing-institute:develop' into add-exception-comments

JimMadge committed:
Fix expected exception message in tests

JimMadge committed:
Change AzureAPIAuthenticationError parent

This prevents allows us to catch this exception specifically and give
useful feedback to the user.

JimMadge commented:
@craddm I think 65b8888 fixes CI.

The issue was making `DataSafeHavenAzureAPIAuthenticationError` a child of `DataSafeHavenAzureError`. We want to distinguish between those so that we can catch API authentication problems and ask the user to run `az login`.

JimMadge committed:
Fix linting

JimMadge commented:
@craddm Is this ready for review?

craddm commented:
Yes, I've decided I'm happy leaving `DSHUserHandlingErrors` alone for the time being, and happy to discuss whether to drop `DSHType/ValueErrors` here

JimMadge commented on data_safe_haven/exceptions/__init__.py:
```diff
@@ -12,61 +18,106 @@ def __init__(self, message: str | bytes):
+    """
+    Exception class for handling errors when authenticating against the Azure API.
+
+    Used to capture exceptions generated when the user is not authenticated or authentication has expired
```
```suggestion
    Used to capture exceptions generated when the user is not authenticated or authentication has expired
```
```suggestion
    Used to capture exceptions generated when the user is not authenticated or authentication has expired.
```

JimMadge commented on data_safe_haven/exceptions/__init__.py:
```diff
@@ -12,61 +18,106 @@ def __init__(self, message: str | bytes):
+class DataSafeHavenMicrosoftGraphError(DataSafeHavenAzureError):
+    """
+    Exception class for handling errors when interacting with the Microsoft Graph API.
+
```
```suggestion
```

JimMadge commented on data_safe_haven/exceptions/__init__.py:
```diff
@@ -12,61 +18,106 @@ def __init__(self, message: str | bytes):
+    """
+    Exception class for handling errors when interacting with Pulumi.
+
+    For example, when a Pulumi operation such as a deployment fails
```
```suggestion
    For example, when a Pulumi operation such as a deployment fails.
```

JimMadge commented on data_safe_haven/exceptions/__init__.py:
```diff
@@ -12,61 +18,106 @@ def __init__(self, message: str | bytes):
+    """
+    Exception class for handling errors related to administration of SSL certificates.
 
+    For example, errors refreshing or creating SSL certificates
```
```suggestion
    For example, errors refreshing or creating SSL certificates.
```

JimMadge commented on data_safe_haven/exceptions/__init__.py:
```diff
@@ -12,61 +18,106 @@ def __init__(self, message: str | bytes):
+    """
+    Exception class for handling errors related to type checking.
 
+    For example, when a function is called with an argument of the wrong type
```
```suggestion
    For example, when a function is called with an argument of the wrong type.
```

JimMadge commented on data_safe_haven/exceptions/__init__.py:
```diff
@@ -12,61 +18,106 @@ def __init__(self, message: str | bytes):
+    """
+    Exception class for handling errors related to user handling.
 
+    For example, when listing or registering users fails
```
```suggestion
    For example, when listing or registering users fails.
```

JimMadge commented on data_safe_haven/exceptions/__init__.py:
```diff
@@ -12,61 +18,106 @@ def __init__(self, message: str | bytes):
+    """
+    Exception class for handling errors related to value checking.
+
+    For example, when a function is called with an argument of the wrong value
```
```suggestion
    For example, when a function is called with an argument of the wrong value.
```

JimMadge reviewed (changes requested):
Looks good.

I'm happy to keep `DataSafeHaven{Value,Type}Error`. Do we catch them anywhere?
If we don't I can see that we might want to in the future.

Looking at commands files, I think there are places where we need to replace `raise DataSafeHaven*` with

```
logger.critical(message)
Typer.Exit(1)
```

@craddm could you open an issue for that unless you want to fix it here?

craddm committed:
Update data_safe_haven/exceptions/__init__.py

Co-authored-by: Jim Madge <jim+github@jmadge.com>

craddm committed:
Update data_safe_haven/exceptions/__init__.py

Co-authored-by: Jim Madge <jim+github@jmadge.com>

craddm committed:
Update data_safe_haven/exceptions/__init__.py

Co-authored-by: Jim Madge <jim+github@jmadge.com>

craddm committed:
Update data_safe_haven/exceptions/__init__.py

Co-authored-by: Jim Madge <jim+github@jmadge.com>

craddm committed:
Update data_safe_haven/exceptions/__init__.py

Co-authored-by: Jim Madge <jim+github@jmadge.com>

craddm committed:
Update data_safe_haven/exceptions/__init__.py

Co-authored-by: Jim Madge <jim+github@jmadge.com>

craddm committed:
Update data_safe_haven/exceptions/__init__.py

Co-authored-by: Jim Madge <jim+github@jmadge.com>

craddm committed:
Use typer.Exit() pattern

JimMadge commented:
@craddm Are you looking to fix all of the exceptions in commands here?

craddm commented:
Possibly - am I on the right track?

JimMadge commented on data_safe_haven/commands/shm.py:
```diff
@@ -114,4 +116,5 @@ def teardown() -> None:
         pulumi_config.upload(context)
     except DataSafeHavenError as exc:
         msg = "Could not teardown Safe Haven Management environment."
-        raise DataSafeHavenError(msg) from exc
+        logger.critical(msg)
```
Could probably put the little string straight in the `logger.critical()` call here.

The reason we often have

```python
msg = "..."
raise Exception(message)
```

is for linting.

JimMadge commented:
> Possibly - am I on the right track?

Yes, you just need to do something similar for every `@typer.command()`.

craddm committed:
Change to raise typer.Exit(1) pattern

craddm committed:
place message directly in logger call

craddm committed:
fix linting

craddm commented:
I think I've got all the relevant ones now. There are some I've left alone as I think we don't want to exit at the point they occur, but could be wrong.

craddm commented:
One thing I was unsure about was when I needed to decode `response.content` to print it as part of the error messages, done in 5598759b8b5724b006fb2e760afab5586302bf56

JimMadge commented on data_safe_haven/external/api/graph_api.py:
```diff
@@ -770,18 +769,17 @@ def http_delete(self, url: str, **kwargs: Any) -> requests.Response:
+        if requests.codes.OK <= response.status_code < requests.codes.MULTIPLE_CHOICES:
+            return response
+        else:
+            msg = f"Could not execute DELETE request to '{url}'. Response content received: '{response.content.decode()}'."
```
```suggestion
            msg = f"Could not execute DELETE request to '{url}'. Response content received: '{response.text}'."
```

I think this is because `response.content` is a `bytes` object, but the exception expects a `str`. I think `Response.text` is a `str`.
If this works it could be applied across all the places.

JimMadge reviewed (changes requested):
@craddm Are there any more `raise`s in `data_safe_haven/commands/*.py` functions? I think all of those should be replaces with a message and `Typer.Exit` to gracefully end the program.

craddm commented:
> @craddm Are there any more `raise`s in `data_safe_haven/commands/*.py` functions? I think all of those should be replaces with a message and `Typer.Exit` to gracefully end the program.

There are some nested `try` blocks in `shm.py`:

https://github.com/craddm/data-safe-haven/blob/43a9698716d7e77a6badd32b6c834a02ce1e011c/data_safe_haven/commands/shm.py#L91-L119

I interpret this as the inner exception being caught by the outer exception, and thus there's no need to `typer.Exit` the inner exception. Or perhaps the inner exception is redundant?

JimMadge committed:
Replace remaining non-typer, terminating raises

JimMadge committed:
Fix linting

JimMadge commented:
@craddm Yes those are OK because we catch them can raise `typer.Exit()`.
There were a couple of others that would terminate the program ungracefully though.
I think 8caebec should fix that.

JimMadge commented:
@craddm Ready to merge?
